src/test/java/io/aiven/connect/elasticsearch/ElasticsearchSinkFailureTest.java

```diff
@@ -0,0 +1,192 @@
+            return (T) success;
+        }
+
+        @Override
+        public <T extends JestResult> void executeAsync(final Action<T> clientRequest,
+                                                        final JestResultHandler<? super T> jestResultHandler) {
+            jestResultHandler.completed((T) success);
+        }
+
+        @Override
+        public void shutdownClient() {
+
+        }
+
+        @Override
+        public void setServers(final Set<String> servers) {
+
+        }
+    }
+
+    abstract class CustomJestElasticsearchClient extends JestElasticsearchClient {
+        public final AtomicInteger apiCallCounter = new AtomicInteger(0);
+
+        public CustomJestElasticsearchClient(final JestClient client) {
+            super(client);
+        }
+
+        @Override
+        public Version getVersion() {
+            return Version.ES_V7;
+        }
+
+        @Override
+        public JsonObject getMapping(final String index, final String type) {
+            return new JsonObject();
+        }
+    }
+
+    class CustomJestTemporarilyFailingElasticsearchClient extends CustomJestElasticsearchClient {
+        public CustomJestTemporarilyFailingElasticsearchClient(final JestClient client) {
+            super(client);
+        }
+
+        @Override
+        public BulkResponse executeBulk(final BulkRequest bulk) throws IOException {
+            final int numAttempt = apiCallCounter.incrementAndGet();
+
+            if (numAttempt < 3) {
+                return BulkResponse.failure(true, "");
+            }
+
+            return BulkResponse.success();
+        }
+    }
+
+    class CustomJestFailingElasticsearchClient extends CustomJestElasticsearchClient {
+        public CustomJestFailingElasticsearchClient(final JestClient client) {
+            super(client);
+        }
+
+        @Override
+        public BulkResponse executeBulk(final BulkRequest bulk) throws IOException {
+            apiCallCounter.incrementAndGet();
+            return BulkResponse.failure(false, "");
+        }
+    }
+
+    @Test
+    public void testRetryIfRecoverable() {
+        final BulkResult success = new BulkResult(new Gson());
+        success.setSucceeded(true);
+
+
+        final ElasticsearchSinkTask elasticsearchSinkTask = new ElasticsearchSinkTask();
+        final CustomJestElasticsearchClient failingClient =
+            new CustomJestTemporarilyFailingElasticsearchClient(new MockedJestClient(success));
+
+        final Map<String, String> props = getConfigWithFilteringOfInvalidMessages();
+
+        elasticsearchSinkTask.start(props, failingClient);
+        elasticsearchSinkTask.put(Collections.singletonList(new SinkRecord("topic", 0, null, null, null, "foo", 0)));
+        elasticsearchSinkTask.flush(Collections.emptyMap());
+        elasticsearchSinkTask.put(Collections.singletonList(new SinkRecord("topic", 0, null, null, null, "bar", 0)));
+        elasticsearchSinkTask.flush(Collections.emptyMap());
```

Comment:
Do we actually need the second record put-flush here? The connector retries by itself (via config setting), so it's confusing to see what's tested here: the retries triggered internally or retries by flushing more records.

Reply:
it was just to show that the first time we try running all 3 the attempts while for the second one, due to the atomic variable we won't proceed with the `add`.
I'm ok to remove the second flush if you think its not adding value

Reply:
I see. Then we should probably add an assertion after the first flush with the api calls = 3 (i.e. enough to succeed), and then a second one after the second flush with assertion api calls has increased in 1